### PR TITLE
Implemented function to compute asymptotic exponents

### PR DESCRIPTION
**Pull request by enocera**
This PR implements functions to compute the asymptotic exponents in the form of plots and of a table of their expectation values and uncertainties at given x and Q. The definition of the asymptotic exponents is as in Eq. (4) of [arXiv:1604.00024]. Examples are provided; the functions are documented. New PDF bases have been defined to introduce PDF combinations that may be relevant for the computation of the asymptotic exponents.

### CODE REVIEW

File: validphys2/src/validphys/asy_exponents.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
+# -*- coding: utf-8 -*-
+"""
+Tools for computing and plotting effective exponents.
+"""
+from __future__ import generator_stop
```
Comment (Zaharid) [inline validphys2/src/validphys/asy_exponents.py:5]: This is not needed for any current version of Python.

File: validphys2/src/validphys/asy_exponents.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    def get_title(self, parton_name):
+        return fr"$\{self.exponent}_a$ for ${parton_name}$ at {format_number(self.Q, 3)} GeV"
+
+    def get_ylabel(self, parton_name):
+        if self.normalize_to is not None:
+            return "Ratio to {}".format(self.normalize_pdf.label)
+        else:
+            return fr"$\{self.exponent}_a$ for ${parton_name}$"
+
+class AsyExponentBandPlotter(BandPDFPlotter, AsyExponentPlotter):
```
Comment (Zaharid) [inline validphys2/src/validphys/asy_exponents.py:152]: Any reason not to merge this into the class above?

File: validphys2/src/validphys/asy_exponents.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+from reportengine.table import table
+
+from validphys.checks import check_positive, check_pdf_normalize_to, make_argcheck, check_xlimits
+from validphys.core import PDF, FitSpec
+from validphys.pdfbases import check_basis, Basis
+from validphys.pdfplots import BandPDFPlotter, PDFPlotter
+
+import validphys.pdfgrids as pdfgrids
+
+from findiff import FinDiff
```
Comment (Zaharid) [inline validphys2/src/validphys/asy_exponents.py:29]: Is this doing anything non trivial? If it was easy I'd rather have the relevant functionality copied somewhere (e.g. calcutils.py) than adding a dependency.

File: validphys2/src/validphys/asy_exponents.py (hunk 1 above)
Comment (enocera) [reply validphys2/src/validphys/asy_exponents.py]: Removed.

File: validphys2/src/validphys/asy_exponents.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+log = logging.getLogger(__name__)
+
+INTERNAL_LINESTYLE = ['-.', ':']
+INTERNAL_COLOR = plt.rcParams['axes.prop_cycle'].by_key()["color"]
```
[the remark this answers is not in the document]
Comment (enocera) [reply validphys2/src/validphys/asy_exponents.py]: Removed.

File: validphys2/src/validphys/asy_exponents.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+import pandas as pd
+
+from reportengine import collect
+from reportengine.compat import yaml
+from reportengine.figure import figuregen
+from reportengine.floatformatting import format_number, significant_digits, format_error_value_columns
+from reportengine.table import table
+
+from validphys.checks import check_positive, check_pdf_normalize_to, make_argcheck, check_xlimits
+from validphys.core import PDF, FitSpec
```
[the remark this answers is not in the document]
Comment (enocera) [reply validphys2/src/validphys/asy_exponents.py]: Removed.

File: validphys2/src/validphys/asy_exponents.py (hunk 2 above)
Comment (enocera) [reply validphys2/src/validphys/asy_exponents.py]: Merged.

File: validphys2/src/validphys/asy_exponents.py (hunk 3 above)
Comment (enocera) [reply validphys2/src/validphys/asy_exponents.py]: This has now been replaced by` np.gradient`, as suggested in #1605

**Review by Zaharid — APPROVED** [review]

**Pull request merged**